File: kotlin/mergeSort.kt
// マージソート
fun main(args: Array<String>) {
    var arr = arrayOf(148, 651, 124, 638, 567, 435, 185, 413, 841, 35);
    var result = margeSort(arr);
    for(i in result) println(i)
}

fun margeSort(arr: Array<Int>): Array<Int> {
    return m(arr, 0, arr.size-1);
}

fun m(arr: Array<Int>, head: Int, tail: Int): Array<Int> {
    if (head < tail) {
        var mid = Math.floor((head + tail).toDouble() / 2).toInt();
        m(arr, head, mid);
        m(arr, mid + 1, tail);
        
        var l = arr.sliceArray(head..(mid));
        var i = 0;
        var j = mid + 1;
        var k = head;
        while (i < l.size && j <= tail) {
            if (l[i] <= arr[j]) {
                arr[k++] = l[i++];
            } else {
                arr[k++] = arr[j++];
            }
        }
        
        while (i < l.size){
            arr[k++] = l[i++];
        }
    }
    return arr;
}

File: python/mergeSort.py
# マージソート
import math
def margeSort(arr):
    def m(head, tail):
        if head < tail:
            mid = math.floor((head + tail) / 2)
            m(head, mid)
            m(mid + 1, tail)
            
            l = arr[head: mid + 1]
            r = arr[mid+1: tail+1]
            k = head
            while len(l) > 0 and len(r) > 0:
                if l[0] <= r[0]:
                    arr[k] = l.pop(0)
                else:
                    arr[k] = r.pop(0)
                k += 1
            
            while len(l) > 0:
                arr[k] = l.pop(0)
                k += 1
        return arr
    return m(0, len(arr)-1)

arr = [104, 865, 413, 541, 304, 354, 165, 150]
print(margeSort(arr))
